### PR TITLE
DRF Login

### PR DESCRIPTION
**Pull request by M-Muaz-Shoukat**
DRF Login to log user in with Json Web Token

### CODE REVIEW

File: todo_list/settings.py [hunk 1]
```diff
@@ -128,6 +129,20 @@
...
 
+
+REST_FRAMEWORK = {
+    'DEFAULT_AUTHENTICATION_CLASSES': (
+        'rest_framework_simplejwt.authentication.JWTAuthentication',
+    )
+}
+
+SIMPLE_JWT = {
+    "ACCESS_TOKEN_LIFETIME": timedelta(minutes=15),
```
Comment (m-shayan-ahmad) [inline todo_list/settings.py]: After 15 minutes will the user be logged out?

File: todo_list/urls.py [hunk 2]
```diff
@@ -16,13 +16,13 @@
...
 from django.urls import path,include
-from todo_list_app.views import index, login_user,logout_user, RegisterUserView, VerifyUserEmailView
+from todo_list_app.views import index, logout_user, RegisterUserView, VerifyUserEmailView, LoginUserView
 
 urlpatterns = [
     path('todo/', include('todo_list_app.urls')),
     path('admin/', admin.site.urls),
     path('', index, name='index'),
-    path('auth/login', login_user, name='login'),
+    path('auth/login/', LoginUserView.as_view(), name='login'),
```
Comment (m-shayan-ahmad) [inline todo_list/urls.py]:
```suggestion
    path('auth/login', LoginUserView.as_view(), name='login'),
```

File: todo_list_app/serializers.py [hunk 3]
```diff
@@ -27,3 +32,34 @@ def create(self, validated_data):
...
+
+    class Meta:
+        model = User
+        fields = ['email', 'password', 'full_name', 'access_token', 'refresh_token']
+
+    def validate(self, attrs):
+        email = attrs.get('email')
+        password = attrs.get('password')
+        request = self.context.get('request')
+        user = authenticate(request, email=email, password=password)
```
Comment (m-shayan-ahmad) [inline todo_list_app/serializers.py]: A serializer should only serialize and validate the data. Authenticating the user within `validate` is not right. A method or class should only do what it's name suggests.

File: todo_list_app/serializers.py [hunk 4]
```diff
@@ -27,3 +32,34 @@ def create(self, validated_data):
...
+            raise AuthenticationFailed('Invalid credentials try again!')
+
+        user_token = user.tokens()
+
+        return {
+            'email': user.email,
+            'full_name': user.get_full_name,
+            'access_token': str(user_token.get('access')),
+            'refresh_token': str(user_token.get('refresh')),
+        }
```
Comment (m-shayan-ahmad) [inline todo_list_app/serializers.py]: Same here. Creating a response with access tokens should not be here.

File: todo_list_app/views.py [hunk 5]
```diff
@@ -69,6 +65,29 @@ def post(self, request):
...
+                'email': user.email,
+                'full_name': user.get_full_name,
+                'user_id': user_id,
+                'message': 'Email is not verified! An email has been sent to your Email Address Verify it.'
+            }, status=status.HTTP_208_ALREADY_REPORTED)
+        return Response(serializer.data, status=status.HTTP_200_OK)
+
+
+class LoginError(Exception):
+    pass
```
Comment (m-shayan-ahmad) [inline todo_list_app/views.py]: Move all errors below imports.

**Review by m-shayan-ahmad — CHANGES_REQUESTED** [review]

File: todo_list_app/serializers.py (hunk 3 above)
Comment (M-Muaz-Shoukat) [reply todo_list_app/serializers.py:51]: validate is authenticating the user to validate if user exists or not
it is kind of validating the user if it exists or not

File: todo_list/settings.py (hunk 1 above)
Comment (m-shayan-ahmad) [reply todo_list/settings.py]: @M-Muaz-Shoukat this was a question.

**Review by m-shayan-ahmad — APPROVED** [review]

**Pull request merged**